### PR TITLE
feat(auth): implement the sign in screen

### PR DESCRIPTION
**Pull request by rBergCode**
# Summary
Implemented the sign in with password screen so that a user can sign in without Google.
In this PR I also fixed some typos which were in the same code base as I was working on. I also fixed some old bad practices.

# Details
The create account is not implemented yet. You can try to sign in with a fake account which email is
"testing@gmail.com" and password "123123"

I will delete this user as soon as this PR is done. It is not a bad thing for it to exist for now as the user can not create or modify any data.

### CODE REVIEW

File: app/src/main/java/ch/hikemate/app/ui/auth/SignInWithEmailScreen.kt [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+            style = TextStyle(fontWeight = FontWeight.Bold, fontSize = 32.sp),
+            modifier = Modifier.testTag(SignInWithEmailScreen.TEST_TAG_TITLE))
+
+        OutlinedTextField(
+            modifier = Modifier.fillMaxWidth().testTag(SignInWithEmailScreen.TEST_TAG_EMAIL_INPUT),
+            keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Email),
+            colors = inputColors,
+            value = email,
+            onValueChange = { email = it },
+            label = { Text(context.getString(R.string.SIGN_IN_WITH_EMAIL_EMAIL_LABEL)) })
```
Comment (MaelImhof) [inline app/src/main/java/ch/hikemate/app/ui/auth/SignInWithEmailScreen.kt:98]: Nitpick: use the composable function `stringResource` instead.

```suggestion
            label = { Text(stringResource(R.string.SIGN_IN_WITH_EMAIL_EMAIL_LABEL)) })
```

File: app/src/main/java/ch/hikemate/app/ui/auth/SignInWithEmailScreen.kt [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+      verticalArrangement = Arrangement.spacedBy(16.dp)) {
+        BackButton(navigationActions)
+        Text(
+            context.getString(R.string.SIGN_IN_WITH_EMAIL_TITLE),
+            style = TextStyle(fontWeight = FontWeight.Bold, fontSize = 32.sp),
+            modifier = Modifier.testTag(SignInWithEmailScreen.TEST_TAG_TITLE))
+
+        OutlinedTextField(
+            modifier = Modifier.fillMaxWidth().testTag(SignInWithEmailScreen.TEST_TAG_EMAIL_INPUT),
+            keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Email),
```
Comment (MaelImhof) [inline app/src/main/java/ch/hikemate/app/ui/auth/SignInWithEmailScreen.kt:94]: Praise: nicely done, this does so much for the UX !

File: app/src/main/java/ch/hikemate/app/ui/auth/SignInWithEmailScreen.kt [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+            label = { Text(context.getString(R.string.SIGN_IN_WITH_EMAIL_EMAIL_LABEL)) })
+
+        OutlinedTextField(
+            modifier =
+                Modifier.fillMaxWidth().testTag(SignInWithEmailScreen.TEST_TAG_PASSWORD_INPUT),
+            visualTransformation = PasswordVisualTransformation(),
+            colors = inputColors,
+            value = password,
+            onValueChange = { password = it },
+            label = { Text(context.getString(R.string.SIGN_IN_WITH_EMAIL_PASSWORD_LABEL)) })
```
Comment (MaelImhof) [inline app/src/main/java/ch/hikemate/app/ui/auth/SignInWithEmailScreen.kt:107]: Nitpick: use the composable function `stringResource` instead.

```suggestion
            label = { Text(stringResource(R.string.SIGN_IN_WITH_EMAIL_PASSWORD_LABEL)) })
```

File: app/src/main/java/ch/hikemate/app/ui/auth/SignInWithEmailScreen.kt [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+
+        BigButton(
+            modifier =
+                Modifier.fillMaxWidth().testTag(SignInWithEmailScreen.TEST_TAG_SIGN_IN_BUTTON),
+            buttonType = ButtonType.PRIMARY,
+            label = context.getString(R.string.SIGN_IN_WITH_EMAIL_SIGN_IN_BUTTON),
```
Comment (MaelImhof) [inline app/src/main/java/ch/hikemate/app/ui/auth/SignInWithEmailScreen.kt:113]: Nitpick: use the composable function `stringResource` instead.

```suggestion
            label = stringResource(R.string.SIGN_IN_WITH_EMAIL_SIGN_IN_BUTTON),
```

File: app/src/test/java/ch/hikemate/app/authentication/AuthViewModelTest.kt [hunk 5]
```diff
@@ -143,7 +143,7 @@ class AuthViewModelTest {
     // Verify that currentUser is initially null
     assertNull(viewModel.currentUser.first())
 
-    viewModel.signInWithEmailAndPassword("mock@example.com", "password")
+    viewModel.signInWithEmailAndPassword("mock@example.com", "password", {}, {})
```
Comment (MaelImhof) [inline app/src/test/java/ch/hikemate/app/authentication/AuthViewModelTest.kt:146]: Nitpick: you might want to make a test where you check that `onSuccess` or `onFailure` are called properly.

File: app/src/main/java/ch/hikemate/app/ui/auth/SignInWithEmailScreen.kt [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+            contentAlignment = Alignment.Center) {
+              TextButton(
+                  onClick = {
+                    // Navigate to the sign up screen
+                    // navigationActions.navigateTo(Screen.SIGN_UP_WITH_EMAIl)
+                  },
+                  modifier = Modifier.testTag(SignInWithEmailScreen.TEST_TAG_GO_TO_SIGN_UP_BUTTON),
+              ) {
+                Text(
+                    context.getString(R.string.SIGN_IN_WITH_EMAIL_GO_TO_SIGN_UP),
```
Comment (MaelImhof) [inline app/src/main/java/ch/hikemate/app/ui/auth/SignInWithEmailScreen.kt:142]: Nitpick: use the composable function `stringResource` instead.

```suggestion
                    stringResource(R.string.SIGN_IN_WITH_EMAIL_GO_TO_SIGN_UP),
```

File: app/src/main/java/ch/hikemate/app/ui/auth/SignInWithEmailScreen.kt [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+      OutlinedTextFieldDefaults.colors()
+          .copy(
+              focusedLabelColor = primaryColor,
+              focusedIndicatorColor = primaryColor,
+              cursorColor = primaryColor,
+              textSelectionColors =
+                  TextSelectionColors(
+                      handleColor = primaryColor,
+                      backgroundColor = primaryColor,
+                  ))
```
Comment (MaelImhof) [inline app/src/main/java/ch/hikemate/app/ui/auth/SignInWithEmailScreen.kt:70]: Praise : very good adaptation to both light and dark modes ! Apart from the `BackButton` but that's not part of the PR.

File: app/src/androidTest/java/ch/hikemate/app/ui/auth/SignInWithEmailScreenTest.kt [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+  fun clickOnSignInButton() {
+    composeTestRule
+        .onNodeWithTag(SignInWithEmailScreen.TEST_TAG_EMAIL_INPUT)
+        .performTextInput("test@gmail.com")
+    composeTestRule
+        .onNodeWithTag(SignInWithEmailScreen.TEST_TAG_PASSWORD_INPUT)
+        .performTextInput("password")
+    composeTestRule.onNodeWithTag(SignInWithEmailScreen.TEST_TAG_SIGN_IN_BUTTON).performClick()
+
+    verify(authRepository, times(1)).signInWithEmailAndPassword(any(), any(), any(), any())
```
Comment (MaelImhof) [inline app/src/androidTest/java/ch/hikemate/app/ui/auth/SignInWithEmailScreenTest.kt:71]: Suggestion: maybe check the email and password to be the ones you entered here?

File: app/src/main/java/ch/hikemate/app/model/authentication/AuthViewModel.kt [hunk 9]
```diff
@@ -65,10 +65,23 @@ class AuthViewModel(private val repository: AuthRepository) : ViewModel() {
...
+  fun signInWithEmailAndPassword(
+      email: String,
+      password: String,
+      onSuccess: () -> Unit,
+      onErrorAction: (Exception) -> Unit
+  ) {
+    if (email.isEmpty() || password.isEmpty()) {
+      onErrorAction(Exception("Email and password must not be empty"))
+      return
+    }
```
[the remark this answers is not in the document]
Comment (rBergCode) [reply app/src/main/java/ch/hikemate/app/model/authentication/AuthViewModel.kt:77]: I did this as the `signInWithEmailAndPassword` from `FirebaseAuth` makes the app crash if it get an empty email or password. I agree this should be managed in the view (for the localization part) so we should probably need to do some sort of error handling (which will be used project-wide).

I don't think this is in the scope of this task. If you agree, you can create a task in the product backlog and we'll do it next sprint (in the same way as moving all `onFailure` to the view)

File: app/src/main/java/ch/hikemate/app/ui/auth/SignInScreen.kt [hunk 10]
```diff
@@ -100,34 +122,41 @@ fun SignInScreen(navigaionActions: NavigationActions) {
...
+) {
   Button(
       onClick = onSignInClick,
       colors = ButtonDefaults.buttonColors(containerColor = Color.White),
       shape = RoundedCornerShape(50),
       modifier =
-          Modifier.padding(8.dp)
+          modifier
+              .padding(8.dp)
               .height(48.dp)
```
[the remark this answers is not in the document]
Comment (rBergCode) [reply app/src/main/java/ch/hikemate/app/ui/auth/SignInScreen.kt:144]: As this is only used here (in a composable, used only in this screen, only for 2 buttons) I think it should be okay to keep it as is. What do you think ?

[the remark this answers is not in the document]
Comment (MaelImhof) [reply app/src/main/java/ch/hikemate/app/ui/auth/SignInScreen.kt:144]: You're right, I think it's ok

File: app/src/main/java/ch/hikemate/app/ui/auth/SignInScreen.kt [hunk 11]
```diff
@@ -100,34 +122,41 @@ fun SignInScreen(navigaionActions: NavigationActions) {
...
         Row(
             verticalAlignment = Alignment.CenterVertically,
             horizontalArrangement = Arrangement.Center,
             modifier = Modifier.fillMaxWidth()) {
               // Load the Google logo from resources
               Image(
-                  painter = painterResource(id = R.drawable.google_logo),
-                  contentDescription = "Google Logo",
+                  painter = painterResource(id = icon),
+                  contentDescription = text,
```
[the remark this answers is not in the document]
Comment (MaelImhof) [reply app/src/main/java/ch/hikemate/app/ui/auth/SignInScreen.kt]: My bad. `contentDescription = null` indeed is the way to go for decorative elements.

File: app/src/test/java/ch/hikemate/app/authentication/AuthViewModelTest.kt (hunk 5 above)
Comment (MaelImhof) [reply app/src/test/java/ch/hikemate/app/authentication/AuthViewModelTest.kt]: Your test will succeed if neither `onSuccess` nor `onFailure` is called. Is that OK?

Comment (rBergCode) [reply app/src/test/java/ch/hikemate/app/authentication/AuthViewModelTest.kt]: This shouldn't happen. I think we're good with that

File: app/src/main/java/ch/hikemate/app/model/authentication/AuthViewModel.kt (hunk 9 above)
[the remark this answers is not in the document]
Comment (MaelImhof) [reply app/src/main/java/ch/hikemate/app/model/authentication/AuthViewModel.kt:77]: Good, the app shouldn't have any crashing bugs for M2. However, we indeed need to handle failure in a better way, and I have no idea how we are exactly going to do that right now. I'll create a task in the product backlog.

File: app/src/test/java/ch/hikemate/app/authentication/AuthViewModelTest.kt (hunk 5 above)
Comment (MaelImhof) [reply app/src/test/java/ch/hikemate/app/authentication/AuthViewModelTest.kt]: Could happen that we forget it when we modify the view model

Comment (MaelImhof) [reply app/src/test/java/ch/hikemate/app/authentication/AuthViewModelTest.kt]: My bad, I did not see the check for the user afterwards. Forget I said anything...
